fix: Filename validation should only forbid `create` and `update`

susnux opened the pull request:
* Resolves: https://github.com/nextcloud/server/issues/44963

## Summary

1. Fix error messages to be consistent and follow the design rules (as they are used by our mobile apps)
2. `isForbidden` must only check full filename - this is used for other purposes (backwards compatibility + fix issues)
3. DAV is directly using the storage and skips their validation, similar other places do, they and the frontend check only the permissions of the node. So we need to adjust the permissions (remove `create` and `update`) when the node is places within an invalid node.
4. The view needs to check also the path for invalid nodes but still needs to allow read-only access on invalid names.

### Screen recording

Setup - folder before enabling "Windows support":
![image](https://github.com/user-attachments/assets/2dd14698-2068-4160-8912-a7e20c6c9bfd)

#### Without this ("before")

[Bildschirmaufnahme_20240812_182432.webm](https://github.com/user-attachments/assets/3c02e1da-f7d2-4971-96ca-0c029d376b03)


#### With this ("after")

[Bildschirmaufnahme_20240812_182149.webm](https://github.com/user-attachments/assets/df64c38f-b86a-4001-b206-e70b095dfd21)


## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc commented:
Hum, so the New button is greyed out because the name of the folder is invalid, but how can the user understand this?
There should be an error message somewhere  explaining this, no?

come-nc commented on lib/private/Files/View.php:
```diff
@@ -1517,6 +1544,9 @@ public function getDirectoryContent($directory, $mimetype_filter = '', ?\OCP\Fil
 						if ($sharingDisabled) {
 							$rootEntry['permissions'] = $rootEntry['permissions'] & ~\OCP\Constants::PERMISSION_SHARE;
 						}
+						if ($folderIsReadonly) {
+							$rootEntry['permissions'] = $rootEntry['permissions'] & ~(\OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE);
+						}
```
It’s allowed to share a file with an invalid name?

susnux replied:
From the feature requirements yes, everything except writing.
So users should still be able to work with it, but not create new (idea is to not block access to existing).

come-nc commented on lib/private/Files/View.php:
```diff
@@ -1826,28 +1856,44 @@ private function getPartFileInfo(string $path): \OC\Files\FileInfo {
 
+		// Handle filename validation for read-only operations, as we still allow to rename invalid files.
+		if ($readonly) {
+			// There is one exception: Forbidden files like `.htaccess` must not be accessible
```
That brings a subtle difference between `forbidden_filenames` and `forbidden_filename_extensions`/`forbidden_filename_basenames`/`forbidden_filename_characters`, that will need to be documented

susnux commented:
> Hum, so the New button is greyed out because the name of the folder is invalid, but how can the user understand this?

Its greyed out because the permission does not include "CREATE". Did not change the front end here, could be done in a follow up, but fixing the permissions.

susnux commented on lib/private/Files/View.php:
```diff
@@ -1097,98 +1099,108 @@ public function free_space($path = '/') {
+			$parent = dirname($path);
+			$parentName = basename($parent);
+			$this->verifyPath($parent, basename($path), !in_array('write', $hooks));
+			// If this is a create operation also verify the parent folder name
+			if (in_array('create', $hooks) && $parentName !== '') {
+				$this->verifyPath(dirname($parent), $parentName);
 			}
```
Yes the path is verified recursively (line 1880)

come-nc replied:
Ok, I was confused by the fact that we call verifyPath manually on the parent here, but it’s because we do the check for non-readonly this second time.

come-nc reviewed (approved):
Approved as `after` is better than `before`. But we should make the situation more clear in the UI in my opinion, or we’ll get confused users about why they do not have write access to the folder, and will look into ACLs and storages and file permissions before thinking about a forbidden name on a parent folder.

juliushaertl commented:
Test failures seem related, looks like it widely fails on install

susnux committed:
fix(dav): Pre-check if node can be copied or moved in DAV files plugin

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix: Renaming does not need update but delete permissions

Renaming is basically copy + delete (a move), so no need to update permissions.
Especially if the node is in a invalid directory the node should be moveable but not editable.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix: Allow read-only filename validation to allow reading files

Needed to read files with the "Windows compatibility" feature.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix: Also validate parent path in `verifyPath`

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
/compile

AndyScherzinger commented:
/compile /

nextcloud-command committed:
chore(assets): Recompile assets

Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

skjnldsv commented:
I guess we would need the parent folder permissiosn to check for CREATE too ?
Maybe we should adjust the api and do something like this?

![image](https://github.com/user-attachments/assets/a83b14f1-8b50-4c59-a2a2-90837d049396)

skjnldsv commented:
> I guess we would need the parent folder permissiosn to check for CREATE too ? Maybe we should adjust the api and do something like this?
> 
> ![image](https://private-user-images.githubusercontent.com/14975046/384291782-a83b14f1-8b50-4c59-a2a2-90837d049396.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwODg0MjgsIm5iZiI6MTczMjA4ODEyOCwicGF0aCI6Ii8xNDk3NTA0Ni8zODQyOTE3ODItYTgzYjE0ZjEtOGI1MC00YzU5LWEyYTItOTA4MzdkMDQ5Mzk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDA3MzUyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3MTNiMWU3M2UzNDAzMTY5ODUzMzMzMmNiYjhlYjI2YzdiNjE2YzI1MzdhNDgzOGI4MzZhYmEzNjJiZDI5ZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9Oftsn1WxZZRaQCJ8qeYjFrKlsD3cYk1RKbkCmUjvI8)

https://github.com/nextcloud-libraries/nextcloud-files/pull/1124
